src/main/java/me/zeroeightsix/kami/module/modules/player/PacketLogger.kt

```diff
@@ -1,100 +1,305 @@
+                                "message: ${it.packet.message}")
+                        }
+                    }
+                    is CPacketKeepAlive -> {
+                        if (!ignoreKeepAlive) {
+                            add(Type.CLIENT, it.packet.javaClass.simpleName,
+                                "key: ${it.packet.key}")
+                        }
+                    }
+                    else -> {
+                        if (!ignoreUnknown) {
+                            add(Type.CLIENT, it.packet.javaClass.simpleName, "Not Registered in PacketLogger.kt")
+                        }
+                    }
+                }
+                if (logInChat) sendChatMessage(lines.joinToString())
+            }
         }
-    }
 
-    private fun write() {
-        defaultScope.launch(Dispatchers.IO) {
-            try {
-                file.bufferedWriter().use {
-                    lines.forEach(it::write)
-                }
-            } catch (e: Exception) {
-                KamiMod.LOG.error("$chatName Error saving!", e)
+        safeListener<TickEvent.ClientTickEvent> {
+            if (showClientTicks) {
+                if (it.phase == TickEvent.Phase.START) lines.add("Tick Pulse - Realtime: ${sdflog.format(Date())} - Runtime: ${System.currentTimeMillis() - start}ms\n")
+                /* Don't let lines get too big, write periodically to the file */
+                if (player.ticksExisted % 200 == 0 || lines.size >= 1000) write()
             }
+        }
 
-            lines.clear()
+        safeListener<ConnectionEvent.Disconnect> {
+            disable()
         }
     }
 
-    private fun read() {
-        defaultScope.launch(Dispatchers.IO) {
-            lines.clear()
-
-            try {
-                file.bufferedReader().forEachLine {
-                    lines.add(it)
+    private fun write() {
+        try {
+            FileWriter(filename, true).also {
+                for (line in lines) it.write(line)
+                it.close()
+            }
+        } catch (e: Exception) {
+            when (e) {
+                is IOException,
+                is ConcurrentModificationException -> {
+                    KamiMod.LOG.error("$chatName Error saving!")
+                    e.printStackTrace()
                 }
-            } catch (e: Exception) {
-                KamiMod.LOG.error("$chatName Error loading!", e)
+                else -> throw e
             }
         }
+        lines.clear()
     }
 
-    init {
-        clear.consumers.add { _, input ->
-            if (input) {
-                lines.clear()
-                write()
-                MessageSendHelper.sendChatMessage("$chatName Packet log cleared!")
-                false
-            } else {
-                input
-            }
-        }
+    /**
+     * Writes a line to the csv following the format:
+     * from (client or server), packet name, time since start, time since last packet, packet data
+     */
```

Comment:
Please use `@param` annotations to explain what `from: Type` is, along with the other two